### DISCUSSION TITLE
Type of waypoint (via or shaping) in GPX extensions

### DISCUSSION THREAD

**Discussion by WolfgangZ42** [devemux86/cruiser#385] [Q&A] [answered]
In a .gpx file routepoints could be classified by either "type" or "extensions" - Examples:

![2023-11-22 15_07_45-D__Dropbox_Motorradtouren_$cruiser_Workflow_Budweis1Shp_e gpx - Notepad++](https://github.com/devemux86/cruiser/assets/121806238/4ac08517-f367-4950-bc84-ac96d52e123b)      ![2023-11-22 15_07_08-D__Dropbox_Motorradtouren_$cruiser_Workflow_Budweis1Shp gpx - Notepad++](https://github.com/devemux86/cruiser/assets/121806238/b9c8ab70-67c0-4da4-af76-b5c5ee672ad7)

Importing a .gpx file with the left specification works fine, "shaping" or "via" are recognized. But with the right format all waypoints become via point, as when "extensions" are just ignored. By the way, the same procedure in Kurviger 2.2.28 and even in kurviger.de. 
Do you see a chance to take the information out of "extensions" as well?

**Comment by devemux86** [comment]
GPX navigation track specifications exist in this topic:
- https://github.com/devemux86/cruiser/discussions/250

It uses the available `type` tag, so no need for extensions.

What is this extension? What website / app generates it?

**Reply by WolfgangZ42** [reply]
> no need for extensions.

You're right - it's my opinion too. But: I get this output from GPXshaping.de (my favorite for inserting shaping points). I asked already months ago - but Markus refused to change it, because some Garmin functions (or BaseCamp) may be going mad. 

In few weeks I have a presentation in front of a bunch of riders and I'll explain my workflow ([Workflow for Kurviger.pdf](https://github.com/devemux86/cruiser/files/13441931/Workflow.for.Kurviger.pdf)). If "extensions" could examined, this could be simplified. The workflow sounds complicated, but believe me it's done in minutes, right after all planning as last step just before starting the adventure. And for me as Cruiserista a way to avoid any deviation even in case of a re-calculation.

**Reply by devemux86** [reply]
@WolfgangZ42 can you post a gpx file with the 2 representations?

Is it for `rtept` that with import they are converted to waypoints (via / shaping)
and cause a route calculation?

**Reply by WolfgangZ42** [reply]
Sure!
The original file from GPXshaping: [Wasserburg am Inn - LandshutShp.gpx.txt](https://github.com/devemux86/cruiser/files/13443139/Wasserburg.am.Inn.-.LandshutShp.gpx.txt)
My edited version of the same file: [Wasserburg am Inn - LandshutShp_e.gpx.txt](https://github.com/devemux86/cruiser/files/13443154/Wasserburg.am.Inn.-.LandshutShp_e.gpx.txt)

And yes only the "rtept" part is of interest. I like to see the route calculation by BRouter because I see all possible deviations in the last planning phase and can correct them.

**Reply by devemux86** [reply]
@WolfgangZ42 thanks for the samples!

So together with the old:
```xml
<rtept>
  <type>via</type>
  <type>shaping</type>
</rtept>
```
The suggestion is to also support the following?
```xml
<rtept>
  <extensions><trp:ViaPoint/></extensions>
  <extensions><trp:ShapingPoint/></extensions>
</rtept>
```

Also I see we use the `type` in both `wpt` and `rtept` in the normal GPX format.
(not GPX navigation track (offline) format)

**Reply by 0709wiwiwi** [reply]
> You're right - it's my opinion too. But: I get this output from GPXshaping.de (my favorite for inserting shaping points). I asked already months ago - but **Markus refused to change it, because some Garmin functions (or BaseCamp) may be going mad.**

Woow....strange, usually it is Garmin driving their customer users mad and not the other way around ;-). 

Using the edited file with extra addition of via added in type, I don't notice anything abnormal in Garmin Mapsource_Basecamp fast tested import.  Garmin does not take the gpx type into account.
[Wasserburg.am.Inn.-.LandshutShp_e_e.gpx.txt](https://github.com/devemux86/cruiser/files/13443533/Wasserburg.am.Inn.-.LandshutShp_e_e.gpx.txt)

**Reply by WolfgangZ42** [reply]
>  to also support the following?

Yes, please! 🙏

**Reply by WolfgangZ42** [reply]
> Garmin does not take the gpx type into account.

Good to know. Thanks for your tests. 
Whereby "via" isn't the problem, seems to be the default, if no gpx type is found. Now I can try to convince GPXshaping to add gpx type in rtept. Perfect would be if Cruiser can support gpx extensions - may be there are such files from other sources as well.

**Reply by 0709wiwiwi** [reply]
Okido Wolfgang, but first this.
I certainly have no objection to Emux reading the gpx correctly.  

Cruiser uses fairly simple and easy understandable gpx files.
Garmin gpx is more complex due to the use of many extensions.
(Building an extra structure into an already existing one).
See a link to the by particularly very many Garmin extensions. 
https://logiqx.github.io/gpx-ideas/garmin/
So you see, who is making who mad ? 
;-)

**Comment by devemux86** [comment]
Garmin extension `trp:ShapingPoint` is supported in the latest:
- https://github.com/devemux86/cruiser/discussions/390

@WolfgangZ42 can you please test with your gpx files?

**Reply by WolfgangZ42** [reply]
Perfect! Thanks a lot for the quick realisation. 👍🙋‍♂️

**Comment by devemux86** [comment, marked as answer]
Implemented in [Cruiser 3.0.37](https://play.google.com/store/apps/details?id=gr.talent.cruiser).
